lms/djangoapps/courseware/access.py

```diff
@@ -507,14 +538,16 @@ def can_load():
         students to see modules.  If not, views should check the course, so we
         don't have to hit the enrollments table on every module load.
         """
-        return (
-            not descriptor.visible_to_staff_only
+        response = (
+            _visible_to_nonstaff_users(descriptor)
             and _has_group_access(descriptor, user, course_key)
-            and (
-                'detached' in descriptor._class_tags  # pylint: disable=protected-access
+            and
+            (
+                _has_detached_class_tag(descriptor)
                 or _can_access_descriptor_with_start_date(user, descriptor, course_key)
             )
-        ) or _has_staff_access_to_descriptor(user, descriptor, course_key)
+        )
+        return response if response else _has_staff_access_to_descriptor(user, descriptor, course_key)
```

Comment:
You'll probably won't get the error object you want here.  If response is false, you're going to return the staff check response.  After checking the staff access, you'll need to return the previously cached response if it was failure, right?